website/archiver/tasks.py

```diff
@@ -0,0 +1,252 @@
+
+from framework.tasks import app as celery_app
+from framework.tasks.utils import logged
+from framework.exceptions import HTTPError
+
+
+from website.archiver import (
+    ARCHIVER_PENDING,
+    ARCHIVER_CHECKING,
+    ARCHIVER_SUCCESS,
+    ARCHIVER_FAILURE,
+    ARCHIVER_SENDING,
+    ARCHIVER_SENT,
+    ARCHIVER_SIZE_EXCEEDED,
+    ARCHIVER_NETWORK_ERROR,
+    ARCHIVER_UNCAUGHT_ERROR,
+    AggregateStatResult,
+)
+from website.archiver import utils
+from website.archiver.model import ArchiveJob
+from website.archiver import signals as archiver_signals
+
+from website.project import signals as project_signals
+from website import settings
+from website.app import init_addons, do_set_backends
+
+def create_app_context():
+    try:
+        init_addons(settings)
+        do_set_backends(settings)
+    except AssertionError:  # ignore AssertionErrors
+        pass
+
+
+logger = get_task_logger(__name__)
+
+class ArchiverTask(celery.Task):
+    abstract = True
+    max_retries = 0
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        if not kwargs:
+            return
+        job = ArchiveJob.load(kwargs['job_pk'])
+        src, dst, user = job.info()
+        reason = None
+        errors = []
+        if isinstance(exc, ArchiverSizeExceeded):
+            dst.archive_status = ARCHIVER_SIZE_EXCEEDED
+            dst.save()
+            reason = ARCHIVER_SIZE_EXCEEDED
+            errors = exc.result
+        elif isinstance(exc, HTTPError):
+            dst.archive_status = ARCHIVER_NETWORK_ERROR
+            dst.save()
+            reason = ARCHIVER_NETWORK_ERROR
+            errors = dst.archive_job.target_info()
+        else:
+            dst.archive_status = ARCHIVER_UNCAUGHT_ERROR
+            dst.save()
+            reason = ARCHIVER_UNCAUGHT_ERROR
+            errors = [einfo]
+        archiver_signals.archive_fail.send(dst, reason=reason, errors=errors)
+        raise exc
+
+class ArchiverSizeExceeded(Exception):
+
+    def __init__(self, result, *args, **kwargs):
+        super(ArchiverSizeExceeded, self).__init__(*args, **kwargs)
+        self.result = result
+
+@celery_app.task(base=ArchiverTask, name="archiver.stat_addon")
+@logged('stat_addon')
+def stat_addon(addon_short_name, job_pk):
+    """Collect metadata about the file tree of a given addon
+
+    :param addon_short_name: AddonConfig.short_name of the addon to be examined
+    :param src_pk: primary key of node being registered
+    :param dst_pk: primary key of registration node
+    :param user_pk: primary key of registration initatior
+    :return: AggregateStatResult containing file tree metadata
+    """
+    create_app_context()
+    job = ArchiveJob.load(job_pk)
+    src, dst, user = job.info()
+    dst.archive_job.update_target(addon_short_name, ARCHIVER_CHECKING)
+    src_addon = src.get_addon(addon_short_name)
+    try:
+        file_tree = src_addon._get_file_tree(user=user)
+    except HTTPError as e:
```

Comment:
Does `_get_file_tree` ever get called in app context? if not I'd recommend not raise `HTTPErrors`